cpp/src/decisiontree/levelalgo/levelhelper_classifier.cuh

```diff
@@ -265,18 +259,25 @@ void get_best_split_classification(
       //Sparse tree
       SparseTreeNode<T, int> &curr_node =
         sparsetree[sparsesize + sparse_nodeid];
-      curr_node.colid = colselector[split_colidx[nodecnt]];
+      int local_colstart = -1;
+      if (h_colstart != nullptr) local_colstart = h_colstart[nodecnt];
+      curr_node.colid =
+        getQuesColumn(h_colids, local_colstart, Ncols, ncols_sampled,
+                      split_colidx[nodecnt], nodecnt);
       curr_node.quesval = getQuesValue(
         minmax, quantile, nbins, split_colidx[nodecnt], split_binidx[nodecnt],
-        nodecnt, n_nodes, colselector, split_algo);
+        nodecnt, n_nodes, curr_node.colid, split_algo);
+
       curr_node.left_child_id = sparsetree_sz + 2 * nodecnt;
       SparseTreeNode<T, int> leftnode, rightnode;
       leftnode.best_metric_val = bestmetric[0];
       rightnode.best_metric_val = bestmetric[1];
       sparsetree.push_back(leftnode);
       sparsetree.push_back(rightnode);
-      sparse_histstate[curr_node.left_child_id] = besthist_left;
-      sparse_histstate[curr_node.left_child_id + 1] = besthist_right;
+      memcpy(&h_child_hist[2 * nodecnt * n_unique_labels], besthist_left.data(),
```

Comment:
Hmm... old code was simpler... any way of reclaiming the simplicity of using the child id field and just assigning the histogram instead of a memcpy?

Reply:
the old code did double buffering when the best split calculation was on GPU. The double buffering was to make memcpyAsyc work. Check here https://github.com/rapidsai/cuml/blob/8a04f2b41a0fdb6bccc81d2d88240ff80044c2ff/cpp/src/decisiontree/levelalgo/levelhelper_classifier.cuh#L172

Now this extra buffers are gone.